Add Dockerfile

tock203 opened the pull request:
Add Dockerfile which describes dependencies for running the test.
This is useful for me to run the test on unprovisioned remote server.

Commit:
Add Dockerfile

Commit:
Make `docker run lightning-integration -k FooNode` possible

cdecker commented on Dockerfile:
```diff
@@ -0,0 +1,87 @@
+    libdb4.8-dev \
+    libdb4.8++-dev \
+    git
+RUN git clone https://github.com/bitcoin/bitcoin.git \
```
This can be done much simpler by just fetching a tarball and unpacking it in the correct place. There is no added value in compiling it in docker imho.

tock203 replied:
If my understanding is correct, current lightning-integration doesn't work with officially released bitcoind binary.
#17 makes lightning-integration only work with bitcoind HEAD.
so I fetch it, then compile it.

cdecker commented on Dockerfile:
```diff
@@ -0,0 +1,87 @@
+# Install bitcoind
+RUN apt-get update \
+  && apt-get install -y software-properties-common \
+  && add-apt-repository ppa:bitcoin/bitcoin \
```
Why do you add the PPA if you then don't download the compiled package?

tock203 replied:
It contains libdb4.8-dev package, which is a dependency for building bitcoind.

cdecker commented on Dockerfile:
```diff
@@ -0,0 +1,87 @@
+  && git clone https://github.com/cdecker/lightning-integration.git \
+  && pip3 install -r lightning-integration/requirements.txt
+
+WORKDIR /root/lightning-integration
```
Setting workdir twice is not really needed if you instead provide a target clone directory. I'd rather have something like this:

```
git clone https://github.com/cdecker/lightning-integration.git /build
WORKDIR /build
```

tock203 replied:
That's true!
I fixed it in https://github.com/cdecker/lightning-integration/pull/35/commits/1fee81ed9bcd62c8dad5ceb0f74088981c99bb49

cdecker commented on Dockerfile:
```diff
@@ -0,0 +1,87 @@
+
+# Install ptarmigan dependencies
+RUN apt-get install -y \
+  git \
```
Some of these are duplicates and I'd rather join all the dependencies into a single `apt-get` call otherwise we get loads of really expensive layers. Combining them also allows us to clear the `apt` caches once they are installed, further slimming the layers.

cdecker commented on Dockerfile:
```diff
@@ -0,0 +1,87 @@
+  jq \
+  bc
+
+RUN make clients
```
Without calling `make update` we'll always have to rebuild the docker images for each run. The `Makefile` provides targets for all the necessary steps. I suggest having `CMD` below execute `make update clients test` and then copy the test results over to `/output` which we can then bind-mount to a host directory allowing us to collect test runs over time and generate the HTML output.

Commit:
Don't fetch unnecessary .git/ dir

Commit:
Don't set WORKDIR twice

Commit:
Install all dependencies in one apt-get to reduce image size

Commit:
Remove the need to rebuild the image for each run

Commit:
Reduce image size

Commit:
Add VOLUME for collecting test result

Commit:
Make `python cli.py html` work

Commit:
Make click work

Without this commit, click will throw
RuntimeError: Click will abort further execution because Python 3 was configured to use ASCII as encoding for the environment.

Commit:
Make `make test` work

cdecker commented on Dockerfile:
```diff
@@ -21,16 +21,17 @@ RUN apt-get update \
-  && git checkout b641f60425674d737d77abd8c49929d953ea4154 \
+    git \
+    curl
+RUN curl -Lo bitcoin.tar.gz https://github.com/bitcoin/bitcoin/archive/v0.17.0rc1.tar.gz \
```
This still compiles, I'd rather just fetch the binary tarball and use that. You can see an example of how to fetch it, verify it and install it here:

https://github.com/ElementsProject/lightning/blob/8ff4418faaeba46e222fe8b65879f54e44ebfc17/Dockerfile#L24-L39

cdecker commented on Dockerfile:
```diff
@@ -37,7 +37,9 @@ RUN apt-get update \
-    zlib1g-dev
+    zlib1g-dev \
+  && apt-get clean \
+  && apt-get autoremove
```
You can go one step further and delete the `apt` caches like this:

https://github.com/ElementsProject/lightning/blob/ab0fa7a1bd156574dd2568bfdbe70255426e928d/contrib/Dockerfile.builder#L38

tock203 replied:
It looks shiny!
fixed

cdecker commented on Dockerfile:
```diff
@@ -63,5 +63,7 @@ ENV PATH $PATH:$GOPATH/bin
 # eclair
 RUN update-alternatives --set java /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java
 
+VOLUME /root/lightning-integration/output
```
Probably also a good idea to declare `results/` a volume, since that's where the raw `py.test` output is stored and it might help debugging a bunch if we keep it.

tock203 replied:
Did you mean `reports/`?
I added it.

ghost commented:
@cdecker How about https://github.com/cdecker/lightning-integration/pull/35#discussion_r211552710 ?
Is that expected behavior?

cdecker committed:
Merge branch 'master' into docker
